imap/conversations.c

```diff
@@ -2357,15 +2361,57 @@ static int conversations_set_guid(struct conversations_state *state,
     buf_printf(&item, "%d:%u", folder, record->uid);
     const char *base = buf_cstring(&item);
 
-    r = conversations_guid_setitem(state, message_guid_encode(&record->guid),
+    const char *guidrep = message_guid_encode(&record->guid);
+    uint64_t internaldate = TIMESPEC_TO_NANOSEC(&record->internaldate);
+    r = conversations_guid_setitem(state, guidrep,
                                    base, record->cid, record->basecid,
                                    record->system_flags,
                                    record->internal_flags,
-                                   record->internaldate,
+                                   internaldate,
                                    add);
+    if (!r) {
+        struct buf key = BUF_INITIALIZER;
+        size_t datalen = 0;
+        const char *data;
+
+        /* Build J key */
+        buf_setcstr(&key, "J");
+        NANOSEC_TO_JMAPID(&key, internaldate);
+
+        /* Fetch existing record (for ref_count) */
+        r = cyrusdb_fetch(state->db, buf_base(&key), buf_len(&key),
+                          &data, &datalen, &state->txn);
+        if (!r || (add && r == CYRUSDB_NOTFOUND)) {
+            uint32_t ref_count = 0;
+            char *p;
+
+            if (datalen) ref_count = strtoul(data, &p, 16);
```

Comment:
Why are the jmapids refcounted by GUID?

Reply:
Because there could be multiple G records with the same value.  If I have the same message in more than one mailbox, I don't want to remove the J key when I remove one instance of the message.

This popped up with one of the Cass tests.

Reply:
That refcount only seems to be relevant when updating a entry in conversations.db, rather than when reading it, right? In this case, it is really necessary to denormalize the refcount here, or would be OK to lookup if any G record on a top-level mailbox for that guid still exists?

Reply:
Yes, I'm uncertain if adding the refcount here is useful enough to warrant a potential data inconsistency. I'm only challenging that implementation, I'm not insisting on removing it.

Typically, one would denormalize data if it brings enough performance improvement. But it looks to me as if in this particular case, the refcount only is evaluated during a write (already a slower operation in comparison to reads) and it only saves reading one row in the db. In contrast, in jmap_mail.c we now do a read for every `_guid_from_id` call because it's cheap to do so. Why isn't it cheap enough here for the same reasons?

Again, this is more an open discussion, rather than an explicit request to remove the refcount.

Reply:
The lookup for an existing GUID would have to be a foreach loop because we need to ignore any G records that match the one we're writing and any that have mime parts.  Or am I thinking about this wrong?

Reply:
yes, it would need to be a loop and only check for G record without imap part (the same as for guid_from_id, right??). Might be that this is too costly, maybe not, hard to tell without actually seeing this in practice. It's OK to keep refcounting.

Reply:
Turns out that ref-counting was broken because it was incrementing everytime a G record was written, including for flag updates.  I removed the ref-counting and just looked for an existing G record to determine whether to add/delete the J record.